scripts/python/phdf.py

```diff
@@ -0,0 +1,382 @@
+            is not found in the file or the data if found.
+
+            If variable is a vector, each element of the returned
+            numpy array is a vector of that length.  
+
+            Default is to return a flat array of length TotalCells.
+            However if flatten is set to False, a 4D (or 5D if vector)
+            array is returned that has dimensions [NumBlocks, Nz, Ny,
+            Nx] where NumBlocks is the total number of blocks, and Nz,
+            Ny, and Nx are the number of cells in the z, y, and x
+            directions respectively.
+
+      ToLocation(index):
+           Returns the location as an array [ib, bidx, iz, iy, ix]
+           which convets the index into a block, index within that
+           block, and z, y, and x locations within that block
+
+      findIndexInOther(self,otherFile,idx,tol=1e-10):
+        Given an index in my data, find the index in a different file.
+        Cell centers have to match within tol.
+        I've curently implemented a naive algorithm and will improve
+        later
+
+    """
+
+    def GenAuxData(self):
+        """
+        Additional attributes filled in by function GenAuxData():
+                          offset[3]: Offsets for real cells in the block
+             isGhost[CellsPerBlock]: Array of size CellsPerBlock indicating ghost cells
+            BlockIdx[CellsPerBlock]: Converts a cell index to [ib, bidx, iz, iy, ix]
+                                     into [ a block ID, index within that
+                                     block, and z, y, and x locations within that block]
+             BlockBounds[NumBlocks]: Bounds of all the blocks
+        """
+        # flag for ghost cells.
+        #Logic is easier starting with all ghost and unmarking
+        self.offset = np.zeros(3,'i')
+        for i in range(3):
+            if self.MeshBlockSize[i] > 1:
+                self.offset[i] = self.NGhost * self.IncludesGhost
+        xRange = range(self.MeshBlockSize[0])
+        yRange = range(self.MeshBlockSize[1])
+        zRange = range(self.MeshBlockSize[2])
+        xo = [self.offset[0], self.MeshBlockSize[0]-self.offset[0]]
+        yo = [self.offset[1], self.MeshBlockSize[1]-self.offset[1]]
+        zo = [self.offset[2], self.MeshBlockSize[2]-self.offset[2]]
+
+        self.BlockIdx = [None]*self.CellsPerBlock
+        self.isGhost = np.ones(self.CellsPerBlock,dtype=bool)
+        index = 0
+        yMask = False
+        zMask = False
+        for k in zRange:
+            if self.NumDims > 2:
+                zMask = (k<zo[0] or k>=zo[1])
+            for j in yRange:
+                if self.NumDims > 1:
+                    yMask = (j<yo[0] or j>=yo[1])
+                for i in xRange:
+                    xMask = (i<xo[0] or i>=xo[1])
+                    self.isGhost[index] = ( xMask or yMask or zMask )
+                    self.BlockIdx[index] = [k, j, i]
+                    index += 1
```

Comment:
I prefer helper functions to be below the constructor. Unless this is a method that can be called *instead* of a constructor, in which case it should be static.

Reply:
fixed.